base/reduce.jl

```diff
@@ -12,6 +12,9 @@ else
     const SmallUnsigned = Union{UInt8,UInt16,UInt32}
 end
 
+abstract type AbstractBroadcasted end
+const AbstractArrayOrBroadcasted = Union{AbstractArray, AbstractBroadcasted}
```

Comment:
I needed to define this `AbstractBroadcasted` as broadcast.jl is loaded later than reduce.jl.  ATM it's just a hack to make this work, but I wonder if it makes sense to define

```julia
abstract type AbstractIndexable end
const Indexable = Union{AbstractArray, AbstractIndexable}
```

or even

```julia
abstract type Indexable{N} end
abstract type AbstractArray{T,N} <: Indexable{N} end
```

so that it's much easier for downstream projects to use indexing-based `mapreduce` implementation.

Reply:
Having an abstract type or a trait like `Indexable` would also be useful to support `mapreduce` over dimensions with `skipmissing` (https://github.com/JuliaLang/julia/pull/28027). It would avoid duplicating some methods.

Though a trait would probably be better than an abstract type, since it would be more flexible. In particular, an object returned by `skipmissing` could be marked as `Indexable` or not depending on whether it wraps an `Indexable` object (like an array) or another iterable. It would also allow objects that already inherit from a non-indexable abstract type to implement `Indexable` if they want, which wouldn't be possible with an abstract type.

Reply:
If we go with trait direction, one option is to use `IndexStyle`.  Something like this:

```julia
struct NonIndexable <: IndexStyle end
IndexStyle(::Type) = NonIndexable()
```

Reply:
That's a clever idea.  I think the union here is sensible for now — that trait could possibly come later.

Reply:
Differing the whole designing work for the trait makes sense.

Reply:
@vtjnash mentioned there may be another way around this bootstrapping issue.

Reply:
Given that the 1.4 feature freeze is approaching, I support merging the PR with the `Union` for now, and try to use a more general trait later.

FWIW, this PR is giving us incredibly good performance to be able to reuse the pairwise summation algorithm with weights at https://github.com/JuliaStats/StatsBase.jl/issues/518. Great work!

Reply:
Thanks a lot for testing this PR in a real-world scenario!